docs/proposals/scheduling/20220609-resource-reservation.md

```diff
@@ -0,0 +1,234 @@
+	// If specifed, the `TTLSeconds` will be ignored.
+	ExpiredTimestamp metav1.Time `json:"expiredTimestamp,omitempty"`
+	// If specified, allocate resources from the referenced objects.
+	ResourceFrom []ReservationOwner `json:"resourceFrom,omitempty"`
+	// By default, reservation's resources requirements (specified in `template.spec`) sums resources would get eventually (specified in `ResourceFrom`) should be not larger than the free resources of the node.
+	// Once `SkipResourceValidate` is set, this validation will be skipped and then overcommitment is allowed.
+	SkipResourceValidate bool `json:"skipResourceValidate,omitempty"`
+}
+
+type ReservationStatus struct {
+	// The `Phase` indicates whether is reservation is waiting for process (`Pending`), available to allocate (`Available`) or expired to get cleanup (Expired).
+	Phase ReservationPhase `json:"phase,omitempty"`
+	// The `Conditions` indicate the messages of reason why the reservation is still pending.
+	Conditions []ReservationCondition `json:"conditions,omitempty"`
+	// Current resource owners which allocated the reservation resources.
+	Owners []ReservationOwner `json:"owners,omitempty"`
+}
+
+type ReservationOwner struct {
+	// refer to a controller when the field `Controller` is set.
+	metav1.OwnerReference
+	// <unresolved>
+	// The `Namespace` field is needed when Reservation is not namespaced.
+	// </unresolved>
+	Namespace string `json:"namespace,omitempty"`
+}
+
+type ReservationOwnerSelector struct {
+	// Multiple field selectors are ORed.
+	ReservationOwner
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+}
+
+type ReservationPhase string
+
+const (
+	ReservationPhasePending   ReservationPhase = "Pending"
+	ReservationPhaseAvailable ReservationPhase = "Available"
+	ReservationPhaseWaiting   ReservationPhase = "Waiting"
+	ReservationPhaseExpired   ReservationPhase = "Expired"
+)
+
+type ReservationCondition struct {
+	LastProbeTime      metav1.Time `json:"lastProbeTime"`
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+	Reason             string      `json:"reason"`
+	Message            string      `json:"message"`
+}
+```
+
+### Implementation Details
+
+#### Reservation Plugin
+
+A `Reservation` object has its resource requirements like a pod. Ideally, A `Reservation` object should get processed directly by the scheduler like a pod. However, it can require a series of modifications on [scheduling framework](#glossary), losing the compatibility with standard kube-scheduler, kubelet, autoscaler, etc. In the reservation plugin, we fake one *reservation pod* for one `Reservation` inside the scheduler to fulfill general scheduling plugins (noderesources, nodeaffinity, tainttolerations, ...). The scheduling framework can handle `Reservation` objects by processing fake pods in both [scheduling cycle and binding cycle](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#scheduling-cycle-binding-cycle).
```

Comment:
It is better to add a diagram to describe the process.

Reply:
Maybe we need more details about the plugin. Roughly describe what extension points need to be implemented, and what needs to be done in the scheduling framework. Otherwise, the description of the implementation details is still too abstract.

Reply:
More descriptions are added.